projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,182 @@
+	if err != nil {
+		return 0, fmt.Errorf("[2] failed to get a valid delay from retry-after header")
+	}
+
+	// if we reach here we have a retry duration (of type time.Duration)
+	switch {
+	case retryDuration > 5*time.Second:
+		return 0, fmt.Errorf("[2] retry-after is more than 5 seconds")
+	case retryDuration > 0:
+		return retryDuration, nil
+	default:
+		return 0, fmt.Errorf("[2] retry-after condition not found")
+	}
+}
+
+// this is returning a responseBody, a retry duration in seconds, and an error
+// this breaks single responsibility... need to rethink it... suggestions?
+func handleStatusCode(response *http.Response) (string, time.Duration, error) {
+	// handle the various response status codes (extensible for more status codes)
+	switch response.StatusCode {
+	// 200
+	case http.StatusOK:
+		// parse the response body
+		responseBody, err := parseResponseBody(response)
+
+		// if we cannot parse the response body then return an error
+		if err != nil {
+			return "", 0, fmt.Errorf("[1] parseResponseBody failed: %w", err)
+		}
+
+		// otherwise return the parsed response body
+		return responseBody, 0, nil
+
+	// 429
+	case http.StatusTooManyRequests:
+		// get the retry duration
+		retryDuration, err := handleTooManyRequestsResponse(response)
+
+		// if we cannot get a valid retry duration then return an error
+		if err != nil {
+			return "", 0, fmt.Errorf("[1] handleTooManyRequestsResponse failed: %w", err)
+		}
+
+		// return a retry duration
+		return "", retryDuration, nil
+
+	// Other Status Codes
+	default:
+		return "", 0, fmt.Errorf("[1] unhandled response status code: %d", response.StatusCode)
+	}
+}
+
+func makeGetRequest(url string) (string, time.Duration, error) {
+	// try to make the get request
+	response, err := http.Get(url)
+
+	// if the request fails then return an error
+	if err != nil {
+		return "", 0, fmt.Errorf("[0] request failed: %w", err)
+	}
+
+	// we must Close the body to signal to Go that we are done with the Response
+	// so it can release all the allocated resources
+	// use the defer keyword early to close the body regardless of the outcome
+	defer response.Body.Close()
+
+	// send the response to the status code handler
+	responseBody, retryDuration, err := handleStatusCode(response)
+
+	// if there is an error
+	if err != nil {
+		return "", 0, fmt.Errorf("[0] handleStatusCode failed: %w", err)
+	}
+
+	// if there is a retry duration
+	if retryDuration > 0 {
+		return "", retryDuration, nil
+	}
+
+	// return the parsed response body
+	return responseBody, 0, nil
+}
+
+func main() {
+	// loop until a successful response or an unrecoverable error
+	for {
+		response, retryDuration, err := makeGetRequest("http://localhost:8080")
+
+		// if we have an unrecoverable error
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			os.Exit(2)
+		}
+
+		// if we have a retry duration
+		if retryDuration > 0 {
+			fmt.Fprintf(os.Stderr, "Retrying after %v...\n", retryDuration)
```

Comment:
```suggestion
			fmt.Fprintf(os.Stderr, "Will retry after %v...\n", retryDuration)
```

With "Retrying" it's unclear from reading whether the delay has already happened or is about to, whereas "Will retry" makes clear "I'm not attempting the retry yet, I'm sleeping first".